crates/storage/provider/src/change.rs

```diff
@@ -349,12 +354,20 @@ impl BundleStateWithReceipts {
         is_value_known: OriginalValuesKnown,
     ) -> Result<(), DatabaseError> {
         // write receipts
+        let mut bodies_cursor = tx.cursor_read::<tables::BlockBodyIndices>()?;
         let mut receipts_cursor = tx.cursor_write::<tables::Receipts>()?;
-        let mut next_number = receipts_cursor.last()?.map(|(i, _)| i + 1).unwrap_or_default();
-        for block_receipts in self.receipts.into_iter() {
-            for receipt in block_receipts {
-                receipts_cursor.append(next_number, receipt)?;
-                next_number += 1;
+
+        for (idx, mut receipts) in self.receipts.into_iter().enumerate() {
+            if !receipts.is_empty() {
+                let (_, body_indices) = bodies_cursor
+                    .seek_exact(self.first_block + idx as u64)?
+                    .expect("body indices exist");
+
+                for (tx_idx, tx_num) in body_indices.tx_num_range().enumerate() {
+                    if let Some(receipt) = receipts.remove(&tx_idx) {
+                        receipts_cursor.append(tx_num, receipt)?;
+                    }
+                }
```

Comment:
Would do this a little bit differently.

You are iterating over blocks `tx_num_range` and then ask `receipts` if there is that number there, so if receipts are empty, you would query hashmap every time.

Would:
```
let first_tx_index = body_indices.first_tx_num();
let mut receipts = receipts.drain().collect::<Vec<_>>();
receipts.sort() // by tx id
for (id,receipt) in receipts {
      receipts_cursor.append(first_tx_index+id,receipt)?;
}
```